### PR TITLE
Disable loading lookups by default in CompactionTask

### PR DESCRIPTION
**Pull request by Akshat-Jain**
## Description

This PR updates CompactionTask to not load any lookups by default, unless `transformSpec` is present.

If transformSpec is present, we will make the decision based on context values, loading all lookups by default. This is done to ensure backward compatibility since transformSpec can reference lookups.
If transform spec is not present and no context value is passed, we donot load any lookup.

This behavior can be overridden by supplying `lookupLoadingMode` and `lookupsToLoad` in the task context.

Other changes/refactoring:
1. Moved CTX_LOOKUP_LOADING_MODE and CTX_LOOKUPS_TO_LOAD constants from PlannerContext to LookupLoadingSpec, as PlannerContext wasn't available in the indexing module.
2. Moved logic for evaluating the context and creating LookupLoadingSpec based on that to `LookupLoadingSpec#getSpecFromContext`.
3. Updated `Task#getLookupLoadingSpec` to return `LookupLoadingSpec.getSpecFromContext(getContext(), LookupLoadingSpec.ALL)`. The default `ALL` ensures that no existing behavior is broken, unless context has been overridden. Making this change in `Task` interface prevents us from having to make this change in a bunch of tasks that are spawned by CompactionTask.

## Test Plan

1. Tried manual compaction with all 3 partitioning modes (dynamic, range, hash) - validated that no lookups are loaded in the compaction task, as well as all the other spawned tasks.
2. Repeated (1) with overriding task context with combination of `lookupLoadingMode` and `lookupsToLoad` - validated that the lookups in all tasks (compaction + spawned) were loaded as per the overridden context. This would help achieve any use-cases where lookups are needed during compaction.
3. Validated that auto-compaction also doesn't load any lookups by default.
4. Also tried ingestion in all 3 partitioning modes (dynamic, range, hash) to ensure this behavior is unaffected.
5. Validated different types of MSQ queries and ingestion to make sure their behavior is unaffected.

This PR has:

- [x] been self-reviewed.
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java [hunk 1]
```diff
@@ -80,6 +88,34 @@ public ImmutableSet<String> getLookupsToLoad()
     return lookupsToLoad;
   }
 
+  public static LookupLoadingSpec getSpecFromContext(Map<String, Object> context, LookupLoadingSpec defaultSpec)
```
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java:91]:
```suggestion
  public static LookupLoadingSpec createFromContext(Map<String, Object> context, LookupLoadingSpec defaultSpec)
```

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java [hunk 2]
```diff
@@ -249,6 +250,12 @@ public CompactionTask(
     this.segmentProvider = new SegmentProvider(dataSource, this.ioConfig.getInputSpec());
     this.partitionConfigurationManager = new PartitionConfigurationManager(this.tuningConfig);
     this.segmentCacheManagerFactory = segmentCacheManagerFactory;
+
+    // Unless context has been overridden to load lookups differently, we want to load no lookups by default in any task spawned
+    // up by the CompactionTask. We achieve this by populating this info in the context which is passed to the spawned tasks.
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java:255]:
```suggestion
    // By default, do not load any lookups in sub-tasks launched by compaction task
```

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java [hunk 3]
```diff
@@ -249,6 +250,12 @@ public CompactionTask(
...
+    if (context == null || !context.containsKey(LookupLoadingSpec.CTX_LOOKUP_LOADING_MODE)) {
+      addToContext(LookupLoadingSpec.CTX_LOOKUP_LOADING_MODE, LookupLoadingSpec.Mode.NONE.toString());
+    }
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java:258]: Use method `addToContextIfAbsent()`.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java [hunk 4]
```diff
@@ -1522,4 +1529,10 @@ public CompactionTuningConfig withPartitionsSpec(PartitionsSpec partitionsSpec)
       );
     }
   }
+
+  @Override
+  public LookupLoadingSpec getLookupLoadingSpec()
+  {
+    return LookupLoadingSpec.getSpecFromContext(getContext(), LookupLoadingSpec.NONE);
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java:1536]: I wonder if the `index_parallel` supervisor tasks launched by compaction will ever need lookups. I should think not. So, this should probably always return `NONE`.

cc: @cryptoe

File: server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java [hunk 5]
```diff
@@ -39,6 +43,10 @@
  */
 public class LookupLoadingSpec
 {
+
+  public static final String CTX_LOOKUP_LOADING_MODE = "lookupLoadingMode";
+  public static final String CTX_LOOKUPS_TO_LOAD = "lookupsToLoad";
```
Comment (kgyrtkirk) [inline server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java:48]: I see a few context keys here-and-there ; but I think it would be better to: co-locate them somewhere
 (or introduce an infra to be pluggable/checkable/etc)

Could these context keys be moved to `QueryContexts` instead (that class is under `processing`) ?

Comment (kfaraz) [inline server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java:48]: I think these constants should live in `Tasks` class instead.

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/PlannerContext.java [hunk 6]
```diff
@@ -357,7 +355,7 @@ public void addLookupToLoad(String lookupName)
   }
 
   /**
-   * Returns the lookup to load for a given task.
+   * Returns the lookup loading spec for a given task.
```
Comment (kfaraz) [inline sql/src/main/java/org/apache/druid/sql/calcite/planner/PlannerContext.java:358]:
```suggestion
   * Lookup loading spec used if this context corresponds to an MSQ task.
```

File: server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java [hunk 7]
```diff
@@ -80,6 +88,34 @@ public ImmutableSet<String> getLookupsToLoad()
...
+
+    if (lookupLoadingMode == LookupLoadingSpec.Mode.NONE) {
+      return LookupLoadingSpec.NONE;
+    } else if (lookupLoadingMode == LookupLoadingSpec.Mode.ALL) {
+      return LookupLoadingSpec.ALL;
+    } else if (lookupLoadingMode == LookupLoadingSpec.Mode.ONLY_REQUIRED) {
+      Collection<String> lookupsToLoad = (Collection<String>) context.get(LookupLoadingSpec.CTX_LOOKUPS_TO_LOAD);
+      if (lookupsToLoad == null || lookupsToLoad.isEmpty()) {
+        throw InvalidInput.exception("Set of lookups to load cannot be %s for mode[ONLY_REQUIRED].", lookupsToLoad);
+      }
```
Comment (kgyrtkirk) [inline server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java:112]: I wonder why its not valid to have it empty - if there are no lookups used; the `ONLY_REQUIRED` should be harmless or not?

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/Task.java [hunk 8]
```diff
@@ -334,9 +334,13 @@ static TaskInfo<TaskIdentifier, TaskStatus> toTaskIdentifierInfo(TaskInfo<Task,
     );
   }
 
+  /**
+   * Fetch lookup loading spec based on the task context, defaulting to loading all lookups if context does not have overridden configuration.
+   * @return lookup loading spec
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/task/Task.java:339]:
```suggestion
   * Specifies the list of lookups to load for this task. Tasks load ALL lookups by default.
   * This behaviour can be overridden by passing parameters {@link Tasks#KEY_LOOKUP_LOADING_MODE}
   * and {@link Tasks#KEY_LOOKUPS_TO_LOAD} in the task context.
```

File: server/src/test/java/org/apache/druid/server/lookup/cache/LookupLoadingSpecTest.java [hunk 9]
```diff
@@ -59,4 +66,62 @@ public void testLoadingOnlyRequiredLookupsWithNullList()
...
+        LookupLoadingSpec.CTX_LOOKUP_LOADING_MODE, LookupLoadingSpec.Mode.ONLY_REQUIRED);
+    final ImmutableMap<String, Object> contextWithModeNone = ImmutableMap.of(
+        LookupLoadingSpec.CTX_LOOKUP_LOADING_MODE, LookupLoadingSpec.Mode.NONE);
+    final ImmutableMap<String, Object> contextWithModeAll = ImmutableMap.of(
+        LookupLoadingSpec.CTX_LOOKUP_LOADING_MODE, LookupLoadingSpec.Mode.ALL);
+    final ImmutableMap<String, Object> contextWithoutLookupKeys = ImmutableMap.of();
+
+    // Return params: <context, default LookupLoadingSpec, expected LookupLoadingSpec>
+    Object[][] params = new Object[][]{
+        // Default spec is returned in the case of context not having the lookup keys.
```
Comment (kgyrtkirk) [inline server/src/test/java/org/apache/druid/server/lookup/cache/LookupLoadingSpecTest.java:93]: I wonder if the intention was to name the set of config somehow; you could use: `Named.of("default",  ... )` to name parameters

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java (hunk 4 above)
Comment (Akshat-Jain) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java:1536]: @kfaraz Based on offline discussion, compaction tasks can technically reference lookups as part of the transformSpec. But that scenario is rarely ever used, so the conclusion was to disable lookup loading by default, while still keeping the option open to be able to override this behavior and load lookups if needed.

Hope that makes sense.

File: server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java (hunk 7 above)
Comment (Akshat-Jain) [reply server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java:112]: Yes, technically it would work. But we decided to keep NONE and ONLY_REQUIRED separate, please check past discussion on this point here: https://github.com/apache/druid/pull/16358#discussion_r1593340894.

Hope this makes sense.

File: server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java (hunk 5 above)
Comment (Akshat-Jain) [reply server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java:48]: @kfaraz `Tasks` class isn't accessible from `server` module. We also need the constants accessible in `LookupLoadingSpec` class for example.
Any other alternate suggestions on where can these be moved?

Comment (Akshat-Jain) [reply server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java:48]: @kgyrtkirk Technically these constants aren't for query contexts, so `QueryContexts` class doesn't seem like a right place to move these to.

File: server/src/test/java/org/apache/druid/server/lookup/cache/LookupLoadingSpecTest.java (hunk 9 above)
Comment (Akshat-Jain) [reply server/src/test/java/org/apache/druid/server/lookup/cache/LookupLoadingSpecTest.java:92]: Sorry, I didn't get it. Could you please elaborate? Thanks!

Not sure if this helps, but the intention here was just to have parameterized test as it would be cleaner than having separate test methods for the different combinations. So `provideParamsForTestCreateFromContext` method is just supplying params to `testGetLookupLoadingSpecFromContext` test via `@MethodSource("provideParamsForTestCreateFromContext")`.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java (hunk 4 above)
Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java]: Yes, it does. 

My question is not about the sub-tasks that will be launched by compaction, rather the supervisor tasks of type `index_parallel`. These supervisor tasks run in the same process as the `CompactionTask` itself and never do any indexing themselves.

So I don't think they would ever need the lookups.

File: server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java (hunk 5 above)
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java:48]: I see, we need these for the new method `createSpecFromContext()`. In that case, you can leave these constants here.

Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java:48]: Yeah, I am not too comfortable with these constants here myself. I had proposed moving them to `Tasks` but that doesn't seem viable either.

I guess we will have to live with them being in `LookupLoadingSpec` itself at least for the time being as there doesn't seem to be a cleaner alternative at the moment.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java (hunk 4 above)
Comment (Akshat-Jain) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java]: @kfaraz I might be misunderstanding your comment due to lack of context (about compaction and supervisors in general), but `index_parallel` is also used for batch ingestion, which can reference lookups. So I'm not sure I understand why they might never need lookups?

Appreciate if you could elaborate, thanks!

File: indexing-service/src/test/java/org/apache/druid/indexing/common/task/ClientCompactionTaskQuerySerdeTest.java [hunk 10]
```diff
@@ -351,7 +358,16 @@ public void testCompactionTaskToClientCompactionTaskQuery() throws IOException
...
-    Assert.assertEquals(expected, actual);
+    Assert.assertEquals(expected.getDataSource(), actual.getDataSource());
+    Assert.assertEquals(expected.getId(), actual.getId());
+    Assert.assertEquals(expected.getType(), actual.getType());
+    Assert.assertEquals(expected.getGranularitySpec(), actual.getGranularitySpec());
+    Assert.assertEquals(expected.getDimensionsSpec(), actual.getDimensionsSpec());
+    Assert.assertEquals(expected.getIoConfig(), actual.getIoConfig());
+    Assert.assertEquals(expected.getTransformSpec(), actual.getTransformSpec());
+    Assert.assertArrayEquals(expected.getMetricsSpec(), actual.getMetricsSpec());
+    Assert.assertEquals(expected.getTuningConfig(), actual.getTuningConfig());
```
Comment (kfaraz) [inline indexing-service/src/test/java/org/apache/druid/indexing/common/task/ClientCompactionTaskQuerySerdeTest.java:369]: Why did we need to break this up? How can we retain the original equality check?

Comment (Akshat-Jain) [reply indexing-service/src/test/java/org/apache/druid/indexing/common/task/ClientCompactionTaskQuerySerdeTest.java:369]: > Why did we need to break this up?

`context={}` vs `context={lookupLoadingMode=NONE}` since we add NONE in the context in CompactionTask's constructor:
```java
    // By default, do not load any lookups in sub-tasks launched by compaction task.
    addToContextIfAbsent(LookupLoadingSpec.CTX_LOOKUP_LOADING_MODE, LookupLoadingSpec.Mode.NONE.toString());
```

> How can we retain the original equality check?

We can't since they're not equal anymore. I could technically pass NONE in the initial context such that `addToContextIfAbsent` doesn't make any difference, and we would be able to retain the original equality check in this test, but that seems like a hacky approach to me. Thoughts?

Comment (kfaraz) [reply indexing-service/src/test/java/org/apache/druid/indexing/common/task/ClientCompactionTaskQuerySerdeTest.java:371]: Hmm, I suppose there is another way. It would not only help us retain the original equality check but also verify that the `CompactionTask` has added something new to it.

```suggestion
    // Verify that CompactionTask has added new parameters into the context 
    Assert.assertNotEquals(expected, actual);
    
    expectedContext.put(LookupLoadingSpec.CTX_LOOKUP_LOADING_MODE, LookupLoadingSpec.Mode.NONE.toString());
    Assert.assertEquals(expected, actual);
```

Comment (kfaraz) [reply indexing-service/src/test/java/org/apache/druid/indexing/common/task/ClientCompactionTaskQuerySerdeTest.java:371]: Create a variable `Map<String, Object> expectedContext = new HashMap<>();`
and use this in the constructor of the `expected` object.

Comment (Akshat-Jain) [reply indexing-service/src/test/java/org/apache/druid/indexing/common/task/ClientCompactionTaskQuerySerdeTest.java]: Done

File: server/src/test/java/org/apache/druid/server/lookup/cache/LookupLoadingSpecTest.java [hunk 11]
```diff
@@ -59,4 +66,67 @@ public void testLoadingOnlyRequiredLookupsWithNullList()
     DruidException exception = Assert.assertThrows(DruidException.class, () -> LookupLoadingSpec.loadOnly(null));
     Assert.assertEquals("Expected non-null set of lookups to load.", exception.getMessage());
   }
+
+  @MethodSource("provideParamsForTestCreateFromContext")
+  @ParameterizedTest
+  public void testGetLookupLoadingSpecFromContext(Map<String, Object> context, LookupLoadingSpec defaultSpec, LookupLoadingSpec expectedSpec)
```
Comment (kfaraz) [inline server/src/test/java/org/apache/druid/server/lookup/cache/LookupLoadingSpecTest.java:72]: This test doesn't really seem like a good candidate for parameterization. Druid tests typically tend to use parameterization when the parameters themselves do not play a role in the assertions. Rather, they tweak some external factor and the assertions must work despite of that factor. For example, using different input formats (csv/json) when verifying that Druid is able to ingest a sample dataset successfully.

In some cases, it might be okay to include the expected value in the parameters but doing that here severely compromises the readability of the test.

Simply put, we should not have to look at the parameter provider method to figure out what the test is even verifying.

---

You could break up this method as appropriate, and each assertion would typically look like:

```java
Assert.assertEquals(
    LookupLoadingSpec.NONE,
    LookupLoadingSpec.createFromContext(
        ImmutableMap.of(LookupLoadingSpec.KEY_1, value1, LookupLoadingSpec.KEY_2, value2),
        LookupLoadingSpec.NONE
    )
);
```

This makes reading through the test much easier.

**Review by kfaraz — COMMENTED** [review]
Left one test related comment, rest of the changes look good.

File: server/src/test/java/org/apache/druid/server/lookup/cache/LookupLoadingSpecTest.java (hunk 11 above)
Comment (Akshat-Jain) [reply server/src/test/java/org/apache/druid/server/lookup/cache/LookupLoadingSpecTest.java]: @kfaraz Have made the change.
Thanks for the detailed info, will keep this in mind! :)

File: server/src/test/java/org/apache/druid/server/lookup/cache/LookupLoadingSpecTest.java [hunk 12]
```diff
@@ -59,4 +61,68 @@ public void testLoadingOnlyRequiredLookupsWithNullList()
...
+    );
+
+    // Default spec is returned in the case of context not having the lookup keys.
+    Assert.assertEquals(
+        LookupLoadingSpec.NONE,
+        LookupLoadingSpec.createFromContext(
+            ImmutableMap.of(),
+            LookupLoadingSpec.NONE
+        )
+    );
```
Comment (kfaraz) [inline server/src/test/java/org/apache/druid/server/lookup/cache/LookupLoadingSpecTest.java:86]: Put the empty context cases in a separate test method.

File: indexing-service/src/test/java/org/apache/druid/indexing/common/task/ClientCompactionTaskQuerySerdeTest.java [hunk 13]
```diff
@@ -71,46 +73,115 @@
...
+                                                       .build();
+  private static final IndexSpec INDEX_SPEC_FOR_INTERMEDIATE_PERSISTS = IndexSpec.builder()
+                                                                                 .withDimensionCompression(CompressionStrategy.LZ4)
+                                                                                 .withMetricCompression(CompressionStrategy.UNCOMPRESSED)
+                                                                                 .withLongEncoding(LongEncodingStrategy.AUTO)
+                                                                                 .build();
+  private static final ClientCompactionIOConfig CLIENT_COMPACTION_IO_CONFIG = new ClientCompactionIOConfig(
+      new ClientCompactionIntervalSpec(
+          Intervals.of("2019/2020"),
+          "testSha256OfSortedSegmentIds"),
```
Comment (kfaraz) [inline indexing-service/src/test/java/org/apache/druid/indexing/common/task/ClientCompactionTaskQuerySerdeTest.java:103]: Can be in a single line.

File: indexing-service/src/test/java/org/apache/druid/indexing/common/task/ClientCompactionTaskQuerySerdeTest.java [hunk 14]
```diff
@@ -71,46 +73,115 @@
...
+  );
+  private static final ClientCompactionTaskGranularitySpec CLIENT_COMPACTION_TASK_GRANULARITY_SPEC =
+      new ClientCompactionTaskGranularitySpec(Granularities.DAY, Granularities.HOUR, true);
+  private static final ClientCompactionTaskDimensionsSpec CLIENT_COMPACTION_TASK_DIMENSIONS_SPEC =
+      new ClientCompactionTaskDimensionsSpec(DimensionsSpec.getDefaultSchemas(ImmutableList.of("ts", "dim")));
+  private static final AggregatorFactory[] METRICS_SPEC = new AggregatorFactory[] {new CountAggregatorFactory("cnt")};
+  private static final ClientCompactionTaskTransformSpec CLIENT_COMPACTION_TASK_TRANSFORM_SPEC =
+      new ClientCompactionTaskTransformSpec(new SelectorDimFilter("dim1", "foo", null));
+  private static final DynamicPartitionsSpec DYNAMIC_PARTITIONS_SPEC = new DynamicPartitionsSpec(100, 30000L);
+  private static final SegmentsSplitHintSpec SEGMENTS_SPLIT_HINT_SPEC = new SegmentsSplitHintSpec(new HumanReadableBytes(100000L), 10);
```
Comment (kfaraz) [inline indexing-service/src/test/java/org/apache/druid/indexing/common/task/ClientCompactionTaskQuerySerdeTest.java:114]: Rather than declaring all of these constants and reusing the same values in all the tests, just add private methods `createCompactionTuningConfig` and `createCompactionTaskQuery`.

Comment (Akshat-Jain) [reply indexing-service/src/test/java/org/apache/druid/indexing/common/task/ClientCompactionTaskQuerySerdeTest.java]: Have created 2 private methods:
1. `createCompactionTaskQuery(String id, ClientCompactionTaskTransformSpec transformSpec)`
2. `createCompactionTask(ClientCompactionTaskTransformSpec transformSpec)`

There are still a bunch of constants left though since they occur in both of the above methods, since the tuning config is different in both of them (ClientCompactionTaskQueryTuningConfig vs TuningConfig). Have removed the constants which didn't occur more than once in the refactored code.

Hope this works.

File: server/src/test/java/org/apache/druid/server/lookup/cache/LookupLoadingSpecTest.java [hunk 15]
```diff
@@ -76,21 +74,44 @@ public void testCreateLookupLoadingSpecFromContext()
...
-    // Default spec is returned in the case of context not having the lookup keys.
     Assert.assertEquals(
         LookupLoadingSpec.NONE,
         LookupLoadingSpec.createFromContext(
             ImmutableMap.of(),
             LookupLoadingSpec.NONE
         )
     );
+  }
+  @Test
```
Comment (kfaraz) [inline server/src/test/java/org/apache/druid/server/lookup/cache/LookupLoadingSpecTest.java:85]:
```suggestion

  @Test
```

**Review by kfaraz — APPROVED** [review]

**Comment by kfaraz** [comment]
Changes look good to me. @kgyrtkirk , @cryptoe , do you have any further feedback?

File: server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java [hunk 16]
```diff
@@ -80,6 +88,34 @@ public ImmutableSet<String> getLookupsToLoad()
...
+    if (context == null) {
+      return defaultSpec;
+    }
+
+    final Object lookupModeValue = context.get(LookupLoadingSpec.CTX_LOOKUP_LOADING_MODE);
+    if (lookupModeValue == null) {
+      return defaultSpec;
+    }
+
+    final LookupLoadingSpec.Mode lookupLoadingMode = LookupLoadingSpec.Mode.valueOf(lookupModeValue.toString());
```
[the remark this answers is not in the document]
Comment (cryptoe) [reply server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java:102]: Error messages should instruct the user what to do.

**Review by cryptoe — COMMENTED** [review]
The description seems outdated. Can you please update @Akshat-Jain

File: server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java [hunk 17]
```diff
@@ -80,6 +88,34 @@ public ImmutableSet<String> getLookupsToLoad()
     return lookupsToLoad;
   }
 
+  public static LookupLoadingSpec createFromContext(Map<String, Object> context, LookupLoadingSpec defaultSpec)
```
[the remark this answers is not in the document]
Comment (Akshat-Jain) [reply server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java:91]: @cryptoe IIUC, the suggestion is to have a `Map<String, Object> defaultContext` as a class variable in LookupLoadingSpec class?

Then the clients would have to update defaultContext in most cases, which seems unnecessary.

Another option could be to add another method:
```java
public static LookupLoadingSpec createFromContext(Map<String, Object> context) {
    return createFromContext(context, LookupLoadingSpec.ALL);
}
```

Thoughts?

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java [hunk 18]
```diff
@@ -249,6 +250,14 @@ public CompactionTask(
...
     this.partitionConfigurationManager = new PartitionConfigurationManager(this.tuningConfig);
     this.segmentCacheManagerFactory = segmentCacheManagerFactory;
+
+    // Do not load any lookups in sub-tasks launched by compaction task, unless transformSpec is present.
+    // If transformSpec is present, we will not modify the context so that the sub-tasks can make the
+    // decision based on context values, loading all lookups by default.
+    // This is done to ensure backward compatibility since transformSpec can reference lookups.
+    if (transformSpec == null) {
+      addToContextIfAbsent(LookupLoadingSpec.CTX_LOOKUP_LOADING_MODE, LookupLoadingSpec.Mode.NONE.toString());
+    }
```
[the remark this answers is not in the document]
Comment (Akshat-Jain) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java:260]: @cryptoe That fallback is present in `Task.java`, so adding it here would be redundant.

File: server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java:91]: I think it is okay to pass the `defaultSpec` to this method, because different call sites may want different defaults if not overridden in the context.

It is similar to `QueryContexts.parseAsString(context, key, defaultValue)`.

File: server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (Akshat-Jain) [reply server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java]: Added.

[the remark this answers is not in the document]
Comment (Akshat-Jain) [reply server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java]: Added.

File: server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java [hunk 19]
```diff
@@ -94,19 +95,34 @@ public static LookupLoadingSpec createFromContext(Map<String, Object> context, L
...
+      return NONE;
+    } else if (lookupLoadingMode == Mode.ALL) {
+      return ALL;
+    } else if (lookupLoadingMode == Mode.ONLY_REQUIRED) {
+      Collection<String> lookupsToLoad;
+      try {
+        lookupsToLoad = (Collection<String>) context.get(CTX_LOOKUPS_TO_LOAD);
+      }
+      catch (ClassCastException e) {
+        throw InvalidInput.exception("Invalid value of %s[%s]. Please provide a comma-separated list of lookup names.",
```
Comment (cryptoe) [inline server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java:122]: Can you please provide an example here ?

**Review by cryptoe — APPROVED** [review]

File: server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java (hunk 19 above)
Comment (Akshat-Jain) [reply server/src/main/java/org/apache/druid/server/lookup/cache/LookupLoadingSpec.java]: Added.

**Review by kgyrtkirk — APPROVED** [review]

**Comment by gianm** [comment]
This patch failed on the group `cds-task-schema-publish-disabled` because it took too long (6h) and was canceled. I don't see such a failure on previous PRs, and I do see it on a bunch of later PRs (#16457, #16252, #16366). Is it possible this patch did something to make that group take a long longer than it used to? In the past, it has taken about 1.5h.

**Comment by Akshat-Jain** [comment]
The cds check passed on the penultimate commit on this PR (https://github.com/apache/druid/actions/runs/9061792112/job/24906027048), and the latest commit didn't have anything other than updating an error message in a couple of files (https://github.com/apache/druid/pull/16420/commits/e6e92a46c63efc45a1d092b378a34f47f662a865).

Also this PR doesn't touch the cds part at all, hence was merged.

File: server/src/test/java/org/apache/druid/server/lookup/cache/LookupLoadingSpecTest.java (hunk 9 above)
Comment (kgyrtkirk) [reply server/src/test/java/org/apache/druid/server/lookup/cache/LookupLoadingSpecTest.java]: that's great and fine ; but instead of leaving comments about a param's intentions - its better to name it with [Named](https://junit.org/junit5/docs/5.9.1/api/org.junit.jupiter.api/org/junit/jupiter/api/Named.html) ; so a more descriptive name will be shown when the testcase is run - an [SO example is here](https://stackoverflow.com/a/66994980/1525291)

Comment (Akshat-Jain) [reply server/src/test/java/org/apache/druid/server/lookup/cache/LookupLoadingSpecTest.java]: @kgyrtkirk I see.
The above test was refactored into non-parameterized tests, so this should be fine, but will keep that in mind for future code changes.
